protobuf/lib/src/protobuf/pb_list.dart

```diff
@@ -6,310 +6,222 @@ part of protobuf;
+  final CheckFunc<E> _check;
 
-  factory FrozenPbList.from(PbList<E> other) =>
-      FrozenPbList._(other._wrappedList);
+  bool _isReadOnly = false;
 
-  UnsupportedError _unsupported(String method) =>
-      UnsupportedError('Cannot call $method on an unmodifiable list');
+  bool get isFrozen => _isReadOnly;
 
-  @override
-  void operator []=(int index, E value) => throw _unsupported('set');
-  @override
-  set length(int newLength) => throw _unsupported('set length');
-  @override
-  void setAll(int index, Iterable<E> iterable) => throw _unsupported('setAll');
-  @override
-  void add(E? element) => throw _unsupported('add');
-  @override
-  void addAll(Iterable<E> iterable) => throw _unsupported('addAll');
-  @override
-  void insert(int index, E element) => throw _unsupported('insert');
-  @override
-  void insertAll(int index, Iterable<E> iterable) =>
-      throw _unsupported('insertAll');
-  @override
-  bool remove(Object? element) => throw _unsupported('remove');
-  @override
-  void removeWhere(bool Function(E element) test) =>
-      throw _unsupported('removeWhere');
-  @override
-  void retainWhere(bool Function(E element) test) =>
-      throw _unsupported('retainWhere');
-  @override
-  void sort([Comparator<E>? compare]) => throw _unsupported('sort');
-  @override
-  void shuffle([math.Random? random]) => throw _unsupported('shuffle');
-  @override
-  void clear() => throw _unsupported('clear');
-  @override
-  E removeAt(int index) => throw _unsupported('removeAt');
-  @override
-  E removeLast() => throw _unsupported('removeLast');
-  @override
-  void setRange(int start, int end, Iterable<E> iterable,
-          [int skipCount = 0]) =>
-      throw _unsupported('setRange');
-  @override
-  void removeRange(int start, int end) => throw _unsupported('removeRange');
-  @override
-  void replaceRange(int start, int end, Iterable<E> newContents) =>
-      throw _unsupported('replaceRange');
-  @override
-  void fillRange(int start, int end, [E? fill]) =>
-      throw _unsupported('fillRange');
-}
-
-class PbList<E> extends PbListBase<E> {
-  PbList({CheckFunc<E> check = _checkNotNull}) : super._noList(check: check);
-
-  PbList.from(List from) : super._from(from);
+  PbList({check = _checkNotNull})
+      : _wrappedList = <E>[],
+        _check = check;
 
-  @Deprecated('Instead use the default constructor with a check function.'
-      'This constructor will be removed in the next major version.')
-  PbList.forFieldType(int fieldType)
-      : super._noList(check: getCheckFunction(fieldType));
+  PbList.unmodifiable()
+      : _wrappedList = const [],
+        _check = _checkNotNull,
+        _isReadOnly = true;
 
-  /// Freezes the list by converting to [FrozenPbList].
-  FrozenPbList<E> toFrozenPbList() => FrozenPbList<E>.from(this);
+  PbList.from(List from)
+      : _wrappedList = List<E>.from(from),
+        _check = _checkNotNull;
 
   @override
   void add(E element) {
-    check(element);
+    if (_isReadOnly) {
```

Comment:
The use of a helper can help control code size.

I would use two levels: (1) a helper that does the check and is inlined (2) a second helper that throws and is never inlined

```dart
Never _readOnlyError(String methodName) {
  throw UnsuppportedError("'$name' on a read-only list");
}

void _checkModifiable(String methodName) {
  if (_isReadOnly) _readOnlyError(methodName);
}

void add(E element) {
  _checkModifiable('add');
}
```

dart2js has a heuristic to not inline methods that only throw.
For AOT you might need `@pragma`s

There is now only one place where there is the combination of calling a constructor and throwing.